ni_measurementlink_generator/ni_measurementlink_generator/templates/internal/_internal.py

```diff
@@ -0,0 +1,109 @@
+"""Helper classes and functions for MeasurementLink examples."""
+
+import logging
+import pathlib
+from typing import Dict, NamedTuple, TypeVar
+
+import grpc
+from ni_measurementlink_service import session_management
+from ni_measurementlink_service._internal.discovery_client import DiscoveryClient
+from ni_measurementlink_service._internal.stubs.ni.measurementlink.pinmap.v1 import (
+    pin_map_service_pb2,
+    pin_map_service_pb2_grpc,
+)
+from ni_measurementlink_service.measurement.service import GrpcChannelPool
+
+
+class ServiceOptions(NamedTuple):
+    """Service options specified on the command line."""
+
+    use_grpc_device: bool = False
+    grpc_device_address: str = ""
+
+    use_simulation: bool = False
+
+
+T = TypeVar("T")
+
+
+def str_to_enum(mapping: Dict[str, T], value: str) -> T:
+    """Convert a string to an enum (with improved error reporting)."""
+    try:
+        return mapping[value]
+    except KeyError as e:
+        logging.error("Unsupported enum value %s", value)
+        raise grpc.RpcError(
+            grpc.StatusCode.INVALID_ARGUMENT,
+            f'Unsupported enum value "{value}"',
+        ) from e
+
+
+class PinMapClient(object):
+    """Class that communicates with the pin map service."""
+
+    def __init__(self, *, grpc_channel: grpc.Channel):
+        """Initialize pin map client."""
+        self._client: pin_map_service_pb2_grpc.PinMapServiceStub = (
+            pin_map_service_pb2_grpc.PinMapServiceStub(grpc_channel)
+        )
+
+    def update_pin_map(self, pin_map_id: str) -> None:
+        """Update registered pin map contents.
+
+        Create and register a pin map if a pin map resource for the specified pin map id is not
+        found.
+
+        Args:
+            pin_map_id (str): The resource id of the pin map to register as a pin map resource. By
+                convention, the pin map id is the .pinmap file path.
+
+        """
+        pin_map_path = pathlib.Path(pin_map_id)
+        request = pin_map_service_pb2.UpdatePinMapFromXmlRequest(
+            pin_map_id=pin_map_id, pin_map_xml=pin_map_path.read_text(encoding="utf-8")
+        )
+        response: pin_map_service_pb2.PinMap = self._client.UpdatePinMapFromXml(request)
+        assert response.pin_map_id == pin_map_id
+
+
+class GrpcChannelPoolHelper(GrpcChannelPool):
```

Comment:
How about GrpcChannelPoolManager ?

Reply:
We should also consider moving this functionality into GrpcChannelPool.